proxy/envoy/ingress.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoy
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/golang/glog"
+
+	"istio.io/manager/model"
+
+	"k8s.io/client-go/pkg/apis/extensions/v1beta1"
+)
+
+type ingressWatcher struct {
+	agent   Agent
+	mesh    *MeshConfig
+	ingress model.IngressRegistry
+}
+
+// NewWatcher creates a new ingress watcher instance with an agent
+func NewIngressWatcher(ingress model.IngressRegistry, ctl model.Controller, mesh *MeshConfig) (Watcher, error) {
+	out := &ingressWatcher{
+		agent:   NewAgent(mesh.BinaryPath, mesh.ConfigPath),
+		mesh:    mesh,
+		ingress: ingress,
+	}
+
+	if err := ctl.AppendIngressHandler(out.notify); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+func (w *ingressWatcher) notify(ingress *v1beta1.Ingress, event model.Event) {
+	config, err := w.generateConfig()
+	if err != nil {
+		glog.Warningf("Failed to generate Envoy configuration: %v", err)
+		return
+	}
+
+	current := w.agent.ActiveConfig()
+	if reflect.DeepEqual(config, current) {
+		glog.V(2).Info("Configuration is identical, skipping reload")
+		return
+	}
+
+	// TODO: add retry logic
+	if err := w.agent.Reload(config); err != nil {
+		glog.Warningf("Envoy reload error: %v", err)
+		return
+	}
+
+	// Add a short delay to de-risk a potential race condition in envoy hot reload code.
+	// The condition occurs when the active Envoy instance terminates in the middle of
+	// the Reload() function.
+	time.Sleep(256 * time.Millisecond)
+}
+
+func (w *ingressWatcher) generateConfig() (*Config, error) {
```

Comment:
On second thought, lets add a TODO here as a reminder that we need to refactor this code. We could do that in a later PR.
Kind of similar to the IdentityProvider stuff in A8, in principle. What do you think?